pkg/build/admission/jenkinsbootstrapper/admission.go

```diff
@@ -0,0 +1,162 @@
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	coreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	kutilerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	"github.com/openshift/origin/pkg/api/latest"
+	authenticationclient "github.com/openshift/origin/pkg/auth/client"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	jenkinscontroller "github.com/openshift/origin/pkg/build/controller/jenkins"
+	"github.com/openshift/origin/pkg/client"
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	"github.com/openshift/origin/pkg/config/cmd"
+)
+
+func init() {
+	admission.RegisterPlugin("JenkinsBootstrapper", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
+		return NewJenkingsBootstrapper(c.Core()), nil
+	})
+}
+
+type jenkingsBootstrapper struct {
+	*admission.Handler
+
+	privilegedRESTClientConfig restclient.Config
+	serviceClient              coreclient.ServicesGetter
+	openshiftClient            client.Interface
+
+	jenkinsConfig configapi.JenkinsPipelineConfig
+}
+
+// NewJenkingsBootstrapper returns an admission plugin that will create required jenkins resources as the user if they are needed.
+func NewJenkingsBootstrapper(serviceClient coreclient.ServicesGetter) admission.Interface {
+	return &jenkingsBootstrapper{
+		Handler:       admission.NewHandler(admission.Create),
+		serviceClient: serviceClient,
+	}
+}
+
+func (a *jenkingsBootstrapper) Admit(attributes admission.Attributes) error {
+	if a.jenkinsConfig.Enabled != nil && !*a.jenkinsConfig.Enabled {
+		return nil
+	}
+	if len(attributes.GetSubresource()) != 0 {
+		return nil
+	}
+	if attributes.GetResource().GroupResource() != buildapi.Resource("buildconfigs") && attributes.GetResource().GroupResource() != buildapi.Resource("builds") {
+		return nil
+	}
+	if !needsJenkinsTemplate(attributes.GetObject()) {
+		return nil
+	}
+
+	namespace := attributes.GetNamespace()
+	glog.V(4).Infof("Handling %v %s/%s", attributes.GetResource(), namespace, attributes.GetName())
+
+	svcName := a.jenkinsConfig.ServiceName
+	if len(svcName) == 0 {
+		return nil
+	}
+
+	// TODO pull this from a cache.
+	// if its not a not-found error (no error or anything else) ,just return ok an take no action.  hopefully it will work.
+	if _, err := a.serviceClient.Services(namespace).Get(svcName); !kapierrors.IsNotFound(err) {
+		if err == nil {
+			glog.V(4).Infof("The Jenkins Pipeline service %q already exists in project %q", svcName, namespace)
+		} else {
+			glog.V(2).Infof("error getting Jenkins Pipeline service %q already exists in project %q: %v", svcName, namespace, err)
```

Comment:
which is weird when I'm thinking about it... this error means the service failed to create but not because it already exists, but something exploded internally.